src/backend_api.cpp

```diff
@@ -0,0 +1,328 @@
+                                           miopenBackendAttributeType_t attributeType,
+                                           int64_t elementCount,
+                                           void* arrayOfElements)
+{
+    if(mFinalized)
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
+
+    switch(attributeName)
+    {
+    case MIOPEN_ATTR_TENSOR_UNIQUE_ID:
+        if(attributeType == MIOPEN_TYPE_INT64 && elementCount == 1)
+        {
+            mBuilder->setId(*static_cast<int64_t*>(arrayOfElements));
+            return;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+    case MIOPEN_ATTR_TENSOR_DATA_TYPE:
+        if(attributeType == MIOPEN_TYPE_DATA_TYPE && elementCount == 1)
+        {
+            mBuilder->setDataType(*static_cast<miopenDataType_t*>(arrayOfElements));
+            return;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+    case MIOPEN_ATTR_TENSOR_DIMENSIONS:
+        if(attributeType == MIOPEN_TYPE_INT64 && elementCount > 0)
+        {
+            mBuilder->setDim(elementCount, static_cast<int64_t*>(arrayOfElements));
+            return;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+    case MIOPEN_ATTR_TENSOR_STRIDES:
+        if(attributeType == MIOPEN_TYPE_INT64 && elementCount > 0)
+        {
+            mBuilder->setStride(elementCount, static_cast<int64_t*>(arrayOfElements));
+            return;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+    case MIOPEN_ATTR_TENSOR_IS_VIRTUAL:
+        if(attributeType == MIOPEN_TYPE_BOOLEAN && elementCount == 1)
+        {
+            mBuilder->setVirtual(*static_cast<bool*>(arrayOfElements));
+            return;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+    case MIOPEN_ATTR_TENSOR_BYTE_ALIGNMENT:
+    case MIOPEN_ATTR_TENSOR_VECTOR_COUNT:
+    case MIOPEN_ATTR_TENSOR_VECTORIZED_DIMENSION:
+    case MIOPEN_ATTR_TENSOR_RAGGED_OFFSET_DESC: MIOPEN_THROW(miopenStatusNotImplemented);
+
+    default: MIOPEN_THROW(miopenStatusBadParm);
+    }
+}
+
+void BackendTensorDescriptor::finalize()
+{
+    mDescriptor = static_cast<TensorDescriptorEx*>(mBuilder->build());
+
+    delete mBuilder;
+    mBuilder = nullptr;
+
+    mFinalized = true;
+}
+
+void BackendTensorDescriptor::getAttribute(miopenBackendAttributeName_t attributeName,
+                                           miopenBackendAttributeType_t attributeType,
+                                           int64_t requestedElementCount,
+                                           int64_t* elementCount,
+                                           void* arrayOfElements)
+{
+    if(!mFinalized)
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
+
+    const auto& dimensions = mDescriptor->GetLengths();
```

Comment:
minor: move declaration closer to use.

Reply:
these are due to declarations are not allowed inside a switch()

Reply:
You are right. There are existing {} already that can be used.